docs: ZKsync SSO Docs

yuliyaalexiev opened the pull request:
# What 💻
Adding ZKsync SSO docs
# Why ✋
Getting ready to launch this product
# Evidence 📷
<img width="1046" alt="image" src="https://github.com/user-attachments/assets/71ea10cc-952d-46d4-915c-1f2d1c7e378e">


<!--

Thank you for contributing to the ZKsync Docs!

Before submitting the PR, please make sure you do the following:

- Update your PR title to follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- Read the [Contributing Guide](https://github.com/matter-labs/zksync-docs/blob/main/CONTRIBUTING.md).
- Understand our [Code of Conduct](https://github.com/matter-labs/zksync-docs/blob/main/CODE_OF_CONDUCT.md)
- Please delete any unused parts of the template when submitting your PR

-->

StanislavBreadless commented on docs/dev/security/bug-bounty.md:
```diff
@@ -2,56 +2,13 @@
-![bounty-table.png](/bounty-table.png)
-
-For severe cases involving a total loss of funds, we will pay a bounty of $500,000.
+We have a bug bounty program at ImmuneFi that is focused on our smart contracts, ZK-SNARK circuits, web and app. Before submitting a bug, please read the [platform rules](https://immunefi.com/rules/) and description of the [bounty program](https://immunefi.com/bounty/zksync/).
```
`web and app`?

yuliyaalexiev committed:
Shell

yuliyaalexiev committed:
Added Interfaces and arch diagram

yuliyaalexiev committed:
Fixed local link

yuliyaalexiev committed:
docs: fixed link

yuliyaalexiev committed:
docs: linting

yuliyaalexiev committed:
docs: new sequence diagram

JackHamer09 committed:
fix: update sso getting started

JackHamer09 committed:
merge

itsacoyote committed:
Merge branch 'main' into ZKsync-SSO

yuliyaalexiev committed:
docs: updated arch diagram background

yuliyaalexiev committed:
Merge branch 'ZKsync-SSO' of https://github.com/matter-labs/zksync-docs into ZKsync-SSO

dutterbutter committed:
chore(build): debugging image links during build

yuliyaalexiev committed:
docs: updated diagram

MexicanAce commented on content/00.build/30.zksync-sso/10.architecture.md:
```diff
@@ -0,0 +1,51 @@
+
+- **ERC7579Account.** Implements the ERC-7579 standard for modular smart contract accounts, allowing for extensible account functionalities.
+- **AAFactory.** A factory contract used to deploy new user accounts on the chain efficiently.
+- **SessionLimitSessionKeyModule.** Manages session keys with specific spend limits, enhancing security by restricting their scope and usage.
```
Is this still the name of the module, I thought @ly0va updated it?

yuliyaalexiev replied:
Good point, changing this to SessionKey - @ly0va please flag if incorrect

ly0va replied:
It's `SessionKeyValidator` now

yuliyaalexiev replied:
Updated

MexicanAce commented on content/00.build/30.zksync-sso/100.faqs.md:
```diff
@@ -0,0 +1,64 @@
+
+## How secure is ZKsync SSO?
+
+ZKsync SSO uses advanced cryptographic techniques and secure authentication methods like passkeys to ensure that your account remains secure.
```
This feels very market-y, can we be more specific what we mean by "advanced cryptographic techniques"?

yuliyaalexiev replied:
good point, removed

MexicanAce commented on content/00.build/30.zksync-sso/20.features.md:
```diff
@@ -0,0 +1,66 @@
+OAuth Recovery ERC-7579 module enables users to set up an account recovery through OAuth tokens associated with their existing Web2 logins like
+Gmail, Facebook, or X (formerly Twitter). This allows users to regain access by logging in with their familiar social media or email accounts.
+
+## User Dashboard
```
Should we mention this if it's not ready?

yuliyaalexiev replied:
Commented out for now - coming soon :)

MexicanAce commented on content/00.build/30.zksync-sso/50.getting-started.md:
````diff
@@ -0,0 +1,106 @@
+  ::code-group
+
+    ```bash [npm]
+    npm i zksync-account
````
The final package name will be `zksync-sso` right?

yuliyaalexiev replied:
Yes, updated

MexicanAce commented on content/00.build/30.zksync-sso/50.getting-started.md:
```diff
@@ -0,0 +1,106 @@
+    session: {
+      feeLimit: parseEther("0.1"),
+      // Allow transfers to a specific address with a limit of 0.1 ETH
+      transferPolicies: [
```
We should include call policies example also

yuliyaalexiev replied:
@JackHamer09  - are you able to add this?

JackHamer09 replied:
Yes, will add

yuliyaalexiev committed:
docs: addressed feedback

yuliyaalexiev committed:
docs: updated to SessionKeyValidator

ly0va commented on content/00.build/30.zksync-sso/00.index.md:
```diff
@@ -0,0 +1,21 @@
+description: Learn about ZKsync Smart Sign-on.
+---
+
+ZKsync SSO is a smart wallet that simplifies user authentication, session management,
```
Are we sure we want to stick to "smart wallet" phrasing? It's not just a wallet after all, it's a whole suite of things

yuliyaalexiev replied:
I am open to removing this, but I believe it is the simplest way to ground devs into envisioning what they are getting.

ly0va commented on content/00.build/30.zksync-sso/10.architecture.md:
```diff
@@ -0,0 +1,51 @@
+## ZKsync SSO Developer SDK
+
+The **Developer SDK** is a client-side library that developers can integrate into their applications with minimal effort.
+It handles authentication, session management, and spend limit controls,
```
throughout the doc, sessions and spend limits are presented as two separate things, whereas spend limits are just one use case of sessions.

cpb8010 replied:
I updated this so it's more clear that you can have sessions and spend limits. While currently we require sessions to have spend limits, you aren't really required to have spend limits on your sessions!

ly0va commented on content/00.build/30.zksync-sso/10.architecture.md:
```diff
@@ -0,0 +1,51 @@
+
+- **Authentication:** Enables users to authenticate using passkeys, supporting biometric methods like Face ID and fingerprint recognition.
+- **Session and Spend Limit Management:** Allows applications to manage user sessions with specific token limits,
+enhancing security by restricting the scope and usage of session keys.
```
I don't think sessions enhance security neither for the developer nor for the user. Can you explain what exactly you mean here?

cpb8010 replied:
I think the idea is that 'smart' sessions *could be* more secure than 'dumb' sessions. We should be clear that the sessions you made a clearly more secure for users **if** developers configure them correctly!

ly0va commented on content/00.build/30.zksync-sso/10.architecture.md:
```diff
@@ -0,0 +1,51 @@
+**Features:**
+
+- **Authentication:** Enables users to authenticate using passkeys, supporting biometric methods like Face ID and fingerprint recognition.
+- **Session and Spend Limit Management:** Allows applications to manage user sessions with specific token limits,
```
Again, sessions aren't only about token limits

ly0va commented on content/00.build/30.zksync-sso/10.architecture.md:
```diff
@@ -0,0 +1,51 @@
+
+- **Passkey Creation:** Assists in generating and managing user passkeys for authentication.
+- **Session Creation:** Manages user sessions, including session keys and their associated permissions.
+- **Transaction Signatures:** Signs transactions on behalf of the user within the constraints of their session limits.
```
That's not true. Auth server does not participate in signing session transactions.

cpb8010 replied:
I assume you can initiate transactions from within the auth page, but this bullet point is super misleading as is. It should could be one of the below:

* Transaction Viewer: Could (eventually) view transactions performed by this account via the domain
* Transaction Relay: Provides a web-interface for applications to approve transactions without needing to redirect to a separate wallet
* Wallet provider: Replaces the traditional wallet extension with a web-native interface for general balance management

ly0va commented on content/00.build/30.zksync-sso/20.features.md:
```diff
@@ -0,0 +1,68 @@
+## Spend Limits
+
+**Spend limits** define the maximum amount of a specific token that can be spent during a session.
+They are an integral part of the session's policy and provide granular control over:
```
Spend limits are not an integral part of the session policy.
In fact, we don't use them in most of our demos.

Also, spend limits are not limited to only ERC-20 tokens.

ly0va commented on content/00.build/30.zksync-sso/25.interfaces.md:
```diff
@@ -0,0 +1,108 @@
+## Sessions
+
+Each allowed `(address, selector)` pair will have an assigned `FunctionPolicy` with all the applied constraints.
+`FunctionPolicy` will have a list of constraints on its arguments, as well as optional cumulative limits and allowances.
```
This section is taken from an already quite outdated design doc, and should be updated. Should also include descriptions of all structs, their fields and methods that session key validator exposes.

cpb8010 replied:
I updated this to just include the typescript interfaces and enums, which is better for external developers. Including our internal module interfaces isn't important as I totally misunderstood how different the context was for this vs our past interface docs.

ly0va commented on content/00.build/30.zksync-sso/30.transaction-flow.md:
```diff
@@ -0,0 +1,49 @@
+- **Action**: ZKsync SSO presents the permissions and actions to be granted, including:
+  - Spend limits
+  - Session key duration
+  - Custom module installations
```
I don't think this is a good example of what will be presented during session creation.

More like:
- allowed methods to call
- allowed arguments to pass
- allowed targets to transfer funds to
- limits restricting spent funds & passed arguments

ly0va reviewed (dismissed):
There seems to be a misunderstanding of how sessions and spend limits are related and how they work. Left some comments

yuliyaalexiev committed:
docs: addressed feedback

JackHamer09 committed:
fix: update getting started with call policies

JackHamer09 committed:
merge

JackHamer09 committed:
fix: update links

JackHamer09 committed:
fix: tabs

JackHamer09 committed:
fix: tabs

JackHamer09 committed:
fix: revert link changes

yuliyaalexiev committed:
docs: interfaces updates

yuliyaalexiev committed:
docs: removed link to nft demo

yuliyaalexiev committed:
doc: diagram updates

yuliyaalexiev committed:
docs: updated diagrams

yuliyaalexiev committed:
doc: updated txn flow

yuliyaalexiev committed:
docs: updated fees

yuliyaalexiev committed:
docs: updated features

cpb8010 committed:
docs: copy types into docs

Replacing unrelated interface information with typescript types.
Also changes in copy to better reflect the standard compliance.

cpb8010 reviewed (commented):
Responded to some comments where I didn't change the content

yuliyaalexiev committed:
docs: interfaces changes

yuliyaalexiev committed:
docs: added links to nft demo

yuliyaalexiev committed:
docs: some wording improvements

yuliyaalexiev committed:
docs: removed spend limits references

cpb8010 committed:
docs: more session interface details

written to explain each field in the interface and rough ideas
on how to use it to do something useful.
Full examples should go somewhere else.

cpb8010 reviewed (dismissed):
I'm happy with the current state, we'll come back and update this as we start fielding more questions from real users.

ly0va commented on content/00.build/30.zksync-sso/25.interfaces.md:
```diff
@@ -0,0 +1,158 @@
+    transferPolicies?: (
+        target: Address;
+        maxValuePerUse?: bigint;
+        valueLimit?: bigint | Limit;)[];
```
missing curly braces around the transfer policy type?

cpb8010 replied:
updated! I fully expand Limit now instead of referencing it.

ly0va commented on content/00.build/30.zksync-sso/25.interfaces.md:
```diff
@@ -0,0 +1,158 @@
+    name: string,
+    icon?: string,
+  },
+  session?: {
```
A bunch of fields here were renamed in @JackHamer09's PR

cpb8010 replied:
updated!
* contractCalls
* address
* transfers

ly0va commented on content/00.build/30.zksync-sso/40.transaction-fees.md:
```diff
@@ -0,0 +1,13 @@
+**Account Creation and Recovery Setup:** Free for users.<br>
+The first 1 million users can create accounts and set up recovery options without paying transaction fees, as ZKsync sponsors these fees.
+
+**Session Creation:** When creating application sessions, users pay a transaction fee (roughly equivalent to a regular transaction fee on ZKsync Era).
```
"regular transaction fee" is a bit vague, a better wording would be "ERC20 transfer"

ly0va commented on content/00.build/30.zksync-sso/50.getting-started.md:
```diff
@@ -0,0 +1,144 @@
+        // Allow calling specific smart contracts (e.g. ERC20 transfer):
+        callPolicies: [
+          {
+            target: "0xa1cf087DB965Ab02Fb3CFaCe1f5c63935815f044", // Contract address
```
renamings here as well

cpb8010 replied:
updated to address!

ly0va reviewed (commented):
should be good enough after the below issues are fixed.
we will improve and expand this later

yuliyaalexiev committed:
docs: updated transaction fee

cpb8010 committed:
docs: update to latest interface

Added more comments to the top level function

cpb8010 committed:
docs: update getting started to latest interface

MexicanAce commented on content/00.build/30.zksync-sso/50.getting-started.md:
````diff
@@ -0,0 +1,144 @@
+    ```ts [wagmi]
+    import { zksync } from "viem/chains";
+    import { createConfig, connect } from "@wagmi/core";
+    import { zksyncAccountConnector } from "zksync-account/connector";
````
This was updated
```suggestion
    import { zksyncAccountConnector } from "zksync-sso/connector";
```

MexicanAce commented on content/00.build/30.zksync-sso/50.getting-started.md:
````diff
@@ -0,0 +1,144 @@
+    ```ts [web3modal]
+    import { zksync } from "viem/chains";
+    import { defaultWagmiConfig } from "@web3modal/wagmi";
+    import { zksyncAccountConnector } from "zksync-account/connector";
````
This was updated
```suggestion
    import { zksyncAccountConnector } from "zksync-sso/connector";
```

MexicanAce committed:
Update content/00.build/30.zksync-sso/50.getting-started.md

MexicanAce committed:
Update content/00.build/30.zksync-sso/50.getting-started.md

MexicanAce commented on content/00.build/30.zksync-sso/50.getting-started.md:
````diff
@@ -0,0 +1,144 @@
+    ::code-group
+
+    ```ts [wagmi]
+    import { zksync } from "viem/chains";
````
Should we tell people to use `zksyncSepoliaTestnet` instead of mainnet? I'm worried people will copy and paste this and be confused why it's not working

JackHamer09 committed:
fix: update interfaces and move getting started section up

JackHamer09 committed:
fix: update main sso page, add links to interfaces and examples

JackHamer09 committed:
fix: use sepolia testnet instead of mainnet

JackHamer09 committed:
fix: lint

JackHamer09 committed:
fix: update transfer policy interface
